README.md

```diff
@@ -254,11 +254,81 @@ we where able to hide it. In the end he is still on doubt if assignments and sta
+
+Well, what is the downside ? On the course (1986) it is mentioned that a lot of languages are exploring this
+idea and they can be used for a lot of things, but operational systems cant be developed with them (at least at
+the time) because of the "dragging tail problem". What would be this problem ? Basically you cant implement
+iterative algorithms on a normal order language, since arguments are always evaluated only when used the
+call graph must be maintained until the argument is evaluated, generating a huge tail. It is the opposite of
+what tail call elimination does when you use recursion, with normal order it is impossible to do tail
+call elimination since by definition you need to maintain the tail until it is lazily evaluated. At the time
+this was a research topic and I don't understand this enough to know if this is still a limitation of
+normal order languages or if there is already some way to circunvent this dowside.
+
+Anyway the streams part of the course gave me a great deal of insight on how lazy evaluation works, and
+how to do it manually in a applicative order language.
+
+On this part of the course some of the downsides of object orientation are discussed against streaming and
+functional programming styles. It is interesting that at the time this was a hot topic and a lot of research
+was being done, and on the course this is presented as a open question and some tradeoffs are discussed
+about maintaining state and side effects versus purely functional. Although usually purely functional
+algotithms are easier to debug and understand, some problems are harder to express like that (the example
+given on the course is a banking system). On the example both streaming programming and functional programming
+lacks the features required to handle the problem properly, which is two different clients manipulating the
+same joint bank account. One of the open problems mentioned is how to isolate this side effect/state part
+of the problem from the rest of the program which is functional, AFAIK this is solved on Haskell with
+[Monads](https://wiki.haskell.org/Monad) for example.
+
+Right now I still lack competence to compare functional and object oriented approaches, specially since
+I'm not very used to purely functional lannguages. It is pretty obvious that everytime you have something
+purely functional, where the substitution model can be applied to your lisp code, debugging and understanding
+will be simpler, but there is too much nuances because of different problems that only hands on experience
+can provide the proper insights.
+
+In the end the more interesting idea presented is that these are tradeoffs, there is no brainless glorification
+of one idea as the best way to solve all the problems, it seems that today we still stand on the same
+point and considering the tradeoffs and working with both techniques seems necessary instead of
+just choosing one of them as the promissed messiah to solve all our problems.
```

Comment:
can you elaborate on the benefits of lazy eval?

Reply:
@tiago4orion I will try on a different essay...for me it is still hard to explain well..because I dont understand the tradeoffs very well yet...just the basic idea =(.

One thing that is interesting is that my whole "kstream" code could be tossed away if I was on a lisp with lazy eval...just using cons/car/cdr would give the same behavior (no need for lazy evaluate the tail manually). But with some simpler examples it may be better to understand this...also I'm not understanding well the downside of it yet...it is like it makes sense intuitively but I do not truly understand it yet =(